### PR TITLE
add upper bound for iterator

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Add upper bound for iterator to speed up region scan when there are a lot of deleted keys.

### CODE REVIEW

**Comment by siddontang** [comment]
Any benchmark?

**Comment by siddontang** [comment]
Please merge master to fix CI first.

**Comment by zhangjinpeng1987** [comment]
In theory, with upper bound key, iterator will return invalid as soon as it reach first key greater than or equal to the upper bound key, regardless of it is a normal key/value pair or deleted key. And I will bench it.

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 1]
```diff
@@ -324,7 +316,7 @@ mod tests {
         assert_eq!(data.len(), 2);
         assert_eq!(data, &base_data[1..3]);
 
-        let mut iter = snap.iter();
+        let mut iter = snap.iter(None);
```
Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: any test to cover not None key?

**Comment by siddontang** [comment]
Ping @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
@BusyJay @disksing @siddontang PTAL

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 2]
```diff
@@ -49,50 +49,55 @@ impl RegionSnapshot {
         &self.region
     }
 
-    pub fn iter(&self) -> RegionIterator {
-        RegionIterator::new(self.snap.new_iterator(), self.region.clone())
+    pub fn iter(&self, upper_bound: Option<&[u8]>) -> RegionIterator {
+        if upper_bound.is_none() {
```
Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: can use `upper_bound.or_else()` to simplify the code.

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 3]
```diff
@@ -49,50 +49,55 @@ impl RegionSnapshot {
...
+        self.scan_impl(try!(self.iter_cf(cf, Some(end_key))), start_key, f)
     }
 
-    fn scan_impl<F>(&self,
-                    mut it: RegionIterator,
-                    start_key: &[u8],
-                    end_key: &[u8],
-                    f: &mut F)
-                    -> Result<()>
+    fn scan_impl<F>(&self, mut it: RegionIterator, start_key: &[u8], f: &mut F) -> Result<()>
```
Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: do we have a test to check that we can't get the key >= upper bound key.

File: src/storage/engine/raftkv.rs [hunk 4]
```diff
@@ -280,17 +281,32 @@ impl Snapshot for RegionSnapshot {
...
 
     #[allow(needless_lifetimes)]
-    fn iter<'b>(&'b self) -> engine::Result<Box<Cursor + 'b>> {
+    fn iter<'b>(&'b self, upper_bound: Option<&[u8]>) -> engine::Result<Box<Cursor + 'b>> {
         SNAPSHOT_OP_COUNTER_VEC.with_label_values(&["iter", "default"]).inc();
 
-        Ok(box RegionSnapshot::iter(self))
+        if upper_bound.is_none() {
+            Ok(box RegionSnapshot::iter(self,
+                                        Some(keys::enc_end_key(self.get_region()).as_slice())))
```
Comment (BusyJay) [inline src/storage/engine/raftkv.rs]: Why `enc_end_key` here?

File: src/storage/mvcc/reader.rs [hunk 5]
```diff
@@ -179,7 +176,7 @@ impl<'a> MvccReader<'a> {
         where F: Fn(&Lock) -> bool
     {
         if self.lock_cursor.is_none() {
-            self.lock_cursor = Some(try!(self.snapshot.iter_cf(CF_LOCK)));
+            self.lock_cursor = Some(try!(self.snapshot.iter_cf(CF_LOCK, None)));
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: Do we need to add upper bound here?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 6]
```diff
@@ -49,50 +49,55 @@ impl RegionSnapshot {
...
 
-    pub fn iter(&self) -> RegionIterator {
-        RegionIterator::new(self.snap.new_iterator(), self.region.clone())
+    pub fn iter(&self, upper_bound: Option<&[u8]>) -> RegionIterator {
+        if upper_bound.is_none() {
+            RegionIterator::new(self.snap.new_iterator(Some(keys::enc_end_key(self.get_region())
+                                    .as_slice())),
+                                self.region.clone())
+        } else {
+            RegionIterator::new(self.snap.new_iterator(upper_bound), self.region.clone())
```
Comment (disksing) [inline src/raftstore/coprocessor/region_snapshot.rs]: I think `upper_bound` need to be encoded here.

Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: It's better to add a new function in RegionIterator to handle the encode/decode thing.

File: src/storage/mvcc/reader.rs (hunk 5 above)
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader.rs]: No, RegionIterator will add region's end_key as upper bound.

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 7]
```diff
@@ -140,7 +126,19 @@ pub struct RegionIterator<'a> {
...
 #[allow(should_implement_trait)]
 impl<'a> RegionIterator<'a> {
-    pub fn new(iter: DBIterator<'a>, region: Region) -> RegionIterator<'a> {
+    pub fn new(snap: &'a Snapshot,
+               region: Region,
+               upper_bound: Option<&[u8]>,
+               cf: &str)
+               -> RegionIterator<'a> {
+        let encoded_upper =
+            upper_bound.map_or_else(|| keys::enc_end_key(&region), |v| keys::data_key(v));
```
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: Just `keys::data_key` should be ok.

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 8]
```diff
@@ -140,7 +126,19 @@ pub struct RegionIterator<'a> {
 // we use rocksdb's style iterator, so omit the warning.
 #[allow(should_implement_trait)]
 impl<'a> RegionIterator<'a> {
-    pub fn new(iter: DBIterator<'a>, region: Region) -> RegionIterator<'a> {
+    pub fn new(snap: &'a Snapshot,
```
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: Can we use two function `new` and `new_cf` here?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 9]
```diff
@@ -381,6 +384,21 @@ mod tests {
...
+        let mut iter = snap.iter(Some(b"a5"));
+        assert!(iter.seek_to_first());
+        let mut res = vec![];
+        loop {
+            res.push((iter.key().to_vec(), iter.value().to_vec()));
+            if !iter.next() {
+                break;
+            }
+        }
+        let expected_res = vec![(b"a1".to_vec(), b"v1".to_vec()), (b"a3".to_vec(), b"v3".to_vec())];
```
Comment (disksing) [inline src/raftstore/coprocessor/region_snapshot.rs]: You can use `base_data` like L373.

Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/region_snapshot.rs]: No, we can't, we add upper_bound for iterator and can't get all datas.

Comment (disksing) [inline src/raftstore/coprocessor/region_snapshot.rs]: You can slice it like L328 :)

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 10]
```diff
@@ -135,7 +126,28 @@ pub struct RegionIterator<'a> {
 // we use rocksdb's style iterator, so omit the warning.
 #[allow(should_implement_trait)]
 impl<'a> RegionIterator<'a> {
-    pub fn new(iter: DBIterator<'a>, region: Region) -> RegionIterator<'a> {
+    pub fn new(snap: &'a Snapshot,
+               region: Region,
+               upper_bound: Option<&[u8]>)
+               -> RegionIterator<'a> {
+        let encoded_upper = upper_bound.map_or_else(|| keys::enc_end_key(&region), keys::data_key);
```
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: What if upper_bound is ""?

**Comment by disksing** [comment]
LGTM

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 11]
```diff
@@ -157,6 +159,14 @@ impl<'a> RegionIterator<'a> {
         }
     }
 
+    fn adjust_upper_bound(upper_bound: Option<&[u8]>) -> Option<&[u8]> {
```
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: Define as a local function instead.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**